### DISCUSSION TITLE
Digest Support for Custom Datasources

### DISCUSSION THREAD

**Discussion by travisshivers** [renovatebot/renovate#25149] [Suggest an Idea]
### Tell us more.

It would be great if custom datasources supported digests.

For example, if the datasource returned something like:
```json
{
  "releases": [
    {
      "version": "main",
      "digest": "fe5edadd9524d5f3378ef39b7ba0b6224d7cf13e"
    }
  ]
}
```

**Comment by secustor** [comment]
This would make sense when the custom manager supports digest pinning or do you have a current use case too?

**Reply by travisshivers** [reply]
Yes, correct, I am aiming to use this with a custom manager for digest pinning too. The custom manager already seems to support digest pinning, so it's just the datasource that is missing digest support currently.

I've got it partially working in https://github.com/renovatebot/renovate/commit/f9ee4d4d15c0df0c04b70b806edc38cad7094cd1 but without implement the full getDigest function because it seems that not enough of the config is passed to the function to properly get the `defaultRegistryUrlTemplate` unlike `getReleases`.

I am trying to get renovate to work with a build system that expresses repository dependencies similarly to git submodules.

And then with that, it works if I capture and pass along the digest as `newDigest`
```js
  customManagers: [
    {
      customType: "regex",
      fileMatch: ["^myDeps.json$"],
      matchStringsStrategy: "recursive",
      matchStrings: [
        '"(?<depName>[^"]*)":\\s*{\\s*"commit":\\s*"(?<currentDigest>[^"]*)"',
      ],
      datasourceTemplate: "custom.myDatasourceWithDigest",
      "currentValueTemplate": "0000000000000000000000000000000000000000",
      "versioningTemplate": "git",
    },
  ],

  customDatasources: {
    myDatasourceWithDigest: {
      format: "json",
      defaultRegistryUrlTemplate: "http://my.api/{{ packageName }}",
      transformTemplates: [
        '{ "releases": $map(releases, function($v) { { "version": "0000000000000000000000000000000000000000", "newDigest": $v.digest } })[] }',
      ],
    },
  },
```

**Reply by secustor** [reply]
I have created a feature request https://github.com/renovatebot/renovate/issues/25197

**Reply by faenschi** [reply]
Thank you so much @secustor !

**Reply by KUGA2** [reply]
I found this digest in the docs for custom datasource but I really dont know what that is. What is a digest? A hash of a file, right?
What is it good for in a custom datasource?
Will the hash be displayed in the PR?

If you have dependencies noted down as url (including version to be updated) and hash, will a custom manager update both version and digest?

(I think the documentation is a bit lacking details here)
